File: src/main/kotlin/org/glow/proto/compiler.kt
package org.glow.proto

/* TODO: translate from JS
function compiler () {
    const types = typesystem.create()
    const repeat = peg.repeat
            const any = peg.any
            const seq = peg.seq

            // ignorable stuff
            const ws = peg.matchWhile(c => c == ' ' || c == '\t' || c == '\v' || c == '\r' || c == '\n', "whitespace", 1)
    const comment = seq(peg.match('#'), peg.matchWhile(c => c != '\n' && c != '\r', "non-newline"))
    const ignored = repeat(any(comment, ws))

    // building blocks
    const opt = parser => repeat(parser, 0, 1).map(v => v.length ? v[0] : null)
    const listOf = parser => seq(parser, repeat(seq(lit(','), parser))).map(pair => [pair[0], ...pair[1].map(x => x[1])])
    const lit = text => peg.match(text).skipping(ignored)

    const id = peg.seq(peg.matchWhile(c => /[A-Za-z_]/.test(c), "alpha", 1), peg.matchWhile(c => /[A-Za-z_0-9]/.test(c), "alphanumeric"))
    .map(args => args[0] + args[1]).skipping(ignored)

    const number = peg.matchWhile(c => /\d/.test(c), "digit", 1).skipping(ignored).map(arg => parseInt(arg, 10))

    const type_expr = peg.lazy()

    const struct_member = peg.seq(id, lit(':'), type_expr).map(args => {
        return [args[0], args[2]]
    })

    const struct_def = seq(lit('struct'), lit('{'), repeat(struct_member, 0), lit('}')).map(args => {
        return types.struct_type(args[2])
    })

    const builtin_type = any(['u8', 'u16', 'u32', 'i8', 'i16', 'i32', 'vu'].map(t => lit(t))).map(id => types.alias_type(id))

    const generic_type_arg = any(type_expr, number)

    const generic_type = seq(id, lit('['), listOf(generic_type_arg), lit(']')).map(args => {
        const alias = types.alias_type(args[0])
        const generic_args = args[2]
        return types.generic_instance(alias, generic_args)
    })

    const basic_type_expr = any(builtin_type, generic_type, struct_def, id.map(name => types.alias_type(name)))

    peg.bind(type_expr, seq(basic_type_expr, repeat(seq(lit('&'), basic_type_expr).map(args => args[1]))).map(args => {
        if (args[1].length > 0) {
            return types.and_type([args[0], ...args[1]])
        }
        else
            return args[0]
    }))

    // this is type definition
    const type_def = seq(lit('type'), id, lit('='), type_expr).map(args => {
        const name = args[1]
        const type = args[3]
        return types.add_type(name, type)
    })

    const ids = listOf(id)
    const arg = seq(id, lit(':'), type_expr).map(args => [args[0], args[2]])
    const args = listOf(arg)
    const method = seq(any(lit('def'), lit('raw')), id, lit('('), opt(args), lit(')'), opt(seq(lit(':'), type_expr))).map(args => {
        const name = args[1]
        const params = args[3] ? args[3] : []
        const reply = args[5] ? args[5][0][1] : undefined
        return types.method(name, params, reply)
    })

    const methods = repeat(method)
    // this is protocol definition
    const proto_def = seq(lit('proto'), id, opt(seq(lit(':'), ids)), lit('{'), methods, lit('}')).map(args => {
        const name = args[1]
        const parents = args[2] ? args[2][1].map(p => types.alias_type(p)) : []
        const methods = args[4]
        return types.add_type(name, types.protocol_type(parents, methods))
    })

    const proto_module = peg.terminated(any(type_def, proto_def), peg.eof.skipping(ignored))
    return {
        number: number,
        id: id,
        listOf: listOf,
        builtin_type: builtin_type,
        generic_type_arg: generic_type_arg,
        generic_type: generic_type,
        type_expr: type_expr,
                // top-level stuff
        type_def: type_def,
        comment: comment,
        ignored: ignored,
        proto_def: proto_def,
        proto_module: proto_module
    }
}
*/

File: firefly/src/compiler.js
import * as peg from './peg.js'
import * as type from './type.js'

export class ProtoCompiler {
    constructor(ts) {
        this.ts = ts
        
        this.ws = peg.sliceMatching("whitespace", 0, x => /\s/.exec(x))
        
        this.lineComment = peg.match("//").flatMap(
            peg.sliceMatching("text", 0, it => it != '\r' && it != '\n')
        ).map(x => x[1]).rename("line-comment")
        
        this.comment = peg.seq(
            peg.match("/*"),
            peg.repeat("comment-body", peg.sliceMatching("non-star", 0, it => it != '*').flatMap(
                peg.sliceMatching("non-slash", 1, it => it == '*')
            ), 0).map(parts => parts.map(x => x[0] + x[1]).join('')),
            peg.match("/")
        ).map(s => s[1].slice(0, s[1].length-1))
        
        this.ignorable = peg.repeat("comment|line-comment|whitespace", peg.any("comment|whitespace", this.comment, this.lineComment, this.ws), 0)
        
        // building blocks

        this.lit = (text) => peg.match(text).skipping(this.ignorable)

        this.delimited = (separator, p) => p.flatMap(
            peg.repeat("tail", peg.seq(this.lit(separator), p).map(x => x[1]), 0)
        ).map (x => [x[0], ...x[1]])
        
        this.id = peg.sliceMatching(
            "id", 1, x => /[a-zA-Z_]/.exec(x)
        ).flatMap(peg.sliceMatching(
            "id_cont", 0, x => /[a-zA-Z0-9_]/.exec(x))
        ).map(x => x[0] + x[1]).skipping(this.ignorable).rename("id")
        
        this.number = peg.sliceMatching("digit", 1, x => /\d/.exec(x)).skipping(this.ignorable).map(it => parseInt(it))
        
        // various types
        
        this.type = peg.lazy()
        this.structMember = peg.seq(this.id, this.lit(":"), this.type).map(x => {
            return { type: x[2], name: x[0] }
        })
        this.structDef = peg.seq(
            this.lit("struct"),
            this.lit("{"),
            peg.repeat("struct members", this.structMember, 0),
            this.lit("}")
        ).map(x => new type.Struct(x[2]))

        this.builtinType = peg.any("built-in type", this.lit("void"), this.lit("int"), this.lit("bool"), this.lit("byte")).map(x => {
            switch(x) {
                case "void": return type.Unit
                case "int": return type.Int
                case "bool":
                case "byte":
                    return type.Byte
            }
        })

        this.genericTypeArg = peg.any("generic argument", this.type, this.number)

        this.genericType = peg.seq(this.id, this.lit("["), this.delimited(",", this.genericTypeArg), this.lit("]")).map (x =>
            new type.Instance(new type.Alias(x[0]), ...x[2])
        )

        this.alias = this.id.map(x => new type.Alias(x))

        this.basicTypePeg = peg.any("basic type expression", this.builtinType, this.genericType, this.structDef, this.alias)
        
        this.type.init(
            peg.seq(this.basicTypePeg, peg.repeat("type-level '&' expression", peg.seq(this.lit("&"), this.basicTypePeg).map(x => x[1]), 0)
        ).map(([head, tail])=> {
            if(tail.length == 0) return head
            else return new type.And([head, ...tail])
        }))

        this.typeDef = peg.seq(this.lit("type"), this.id, this.lit("="), this.type).map(x => {
            return { name: x[1], type: x[3] }
        }).skipping(this.ignorable)
        
        this.aliases = this.delimited(",", this.alias)
        
        this.arg = peg.seq(this.id, this.lit(":"), this.type).map(x => {
            return { name: x[0], type: x[2] }
        })

        this.args = peg.seq(this.lit("("), peg.optional("args", this.delimited(",", this.arg)), this.lit(")")).map(x => 
            x[1] ? x[1] : []
        )
        
        this.returnType = peg.optional("return type", peg.seq(this.lit(":"), this.type)).map(x => x ? x[1] : type.Unit)
        
        this.method = peg.seq(peg.any("def|msg", this.lit("def"), this.lit("msg")), this.id, this.args, this.returnType).map(([kind, name, args, ret]) => {
            return new type.Method(kind, name, args, ret ? ret : type.Unit)
        })

        this.methods = peg.repeat("methods", this.method, 0)
        
        // this is protocol definition
        this.extendsProtos = peg.optional("extended protos", peg.seq(this.lit(":"), this.aliases)).map( it =>
            it ? it[1] : []
        )

        this.protoBody = peg.seq(this.lit("{"), this.methods, this.lit("}")).map(x => x[1])
       
        this.protoDef = peg.seq(this.lit("proto"), this.id, this.extendsProtos, this.protoBody).map( ([_, name, extended, methods]) => {
            return { name: name, type : new type.Proto(name, extended, methods) }
        }).skipping(this.ignorable)
       
        this.decl = peg.any("declaration", this.typeDef, this.protoDef)
        this.module = peg.repeat("module", this.decl, 0).flatMap(this.ignorable).flatMap(peg.EOF).rename("module").map(([decls, _]) => {
            const members = {}
            for (const decl of decls[0]) {
                members[decl.name] = decl.type
                ts.register(decl.name, decl.type)
            }
            return new type.Module(members)
        })
    }
}